prost-build/src/code_generator.rs

```diff
@@ -676,6 +696,233 @@ impl<'a> CodeGenerator<'a> {
+        self.depth -= 1;
+        self.push_indent();
+        self.buf.push_str("}\n");
+
+        self.depth -= 1;
+        self.push_indent();
+        self.buf.push_str("}\n");
+
+        self.push_indent();
+        self.buf.push_str("impl From<");
+        self.buf.push_str(builder_name);
+        self.buf.push_str("> for super::");
+        self.buf.push_str(&struct_name);
+        self.buf.push_str(" {\n");
+        self.depth += 1;
+        self.push_indent();
+        self.buf.push_str("#[inline]\n");
+        self.push_indent();
+        self.buf.push_str("fn from(builder: ");
+        self.buf.push_str(builder_name);
+        self.buf.push_str(") -> Self {\n");
+        self.depth += 1;
+        self.push_indent();
+        self.buf.push_str("builder.inner\n");
+        self.depth -= 1;
+        self.push_indent();
+        self.buf.push_str("}\n");
+        self.depth -= 1;
+        self.push_indent();
+        self.buf.push_str("}\n");
+    }
+
+    fn append_builder_field_setter(&mut self, fq_message_name: &str, field: &Field) {
+        let rust_name = field.rust_name();
+        let repeated = field.descriptor.label == Some(Label::Repeated as i32);
+        let deprecated = self.deprecated(&field.descriptor);
+        let optional = self.optional(&field.descriptor);
+        let boxed = self.boxed(&field.descriptor, fq_message_name, None);
+        let ty = self.resolve_type(&field.descriptor, fq_message_name);
+
+        let prost_path = prost_path(self.config);
+        let arg_type = if field.descriptor.r#type() == Type::Enum {
+            // Enums are special: the field type is i32, but for the setter
+            // we want the accompanying enumeration type.
+            self.resolve_ident(field.descriptor.type_name())
+        } else {
+            let arg_bound = if boxed {
+                format!("Into<{}::alloc::boxed::Box<{}>>", prost_path, ty)
+            } else {
+                format!("Into<{}>", ty)
+            };
+            if repeated {
+                format!("impl IntoIterator<Item = impl {}>", arg_bound)
+            } else {
+                format!("impl {}", arg_bound)
+            }
+        };
+
+        debug!("    field setter: {:?}, arg: {:?}", rust_name, arg_type);
+
+        if deprecated {
+            self.push_indent();
+            self.buf.push_str("#[deprecated]\n");
+        }
+        self.push_indent();
+        self.buf.push_str("#[inline]\n");
+        self.push_indent();
+        self.buf.push_str("pub fn ");
+        self.buf.push_str(&rust_name);
+        self.buf.push_str("(mut self, value: ");
+        self.buf.push_str(&arg_type);
+        self.buf.push_str(") -> Self {\n");
+        self.depth += 1;
+
+        self.push_indent();
+        self.buf.push_str("self.inner.");
+        self.buf.push_str(&rust_name);
+        self.buf.push_str(" = ");
+        if repeated {
+            self.buf
+                .push_str("value.into_iter().map(Into::into).collect();\n");
```

Comment:
A repeated type is a `Vec`, but this setter doesn't allow the direct setting of a `Vec`. It will reallocate. I think the setter should just use `Vec` as its input type.

Reply:
I consider the setter a convenience method. The field is public, so there is always an option of setting it directly rather than going through `FromIterator`.

When specialization is stabilized in future Rust, the setter can be specialized for `Vec` without breaking the API.

Reply:
An `impl Into<Vec<T>>` argument looks like a reasonable compromise: apart from `Vec<T>` with zero copying,  it accepts `&[T]`, `&[T; N]`, and `[T; N]`, so it should cover the most common cases without extra conversions. To set from an iterator, the caller would have to call `.collect::<Vec<_>>()`